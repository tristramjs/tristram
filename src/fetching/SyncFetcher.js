/* @flow */
import type { RawSiteMapData } from '../types/sitemap';

import type { ChunkFetcher } from './index';

export default class SyncFetcher implements ChunkFetcher {
	data: RawSiteMapData | RawSiteMapData[];

	constructor({ data }: { data: RawSiteMapData | RawSiteMapData[] }) {
		this.data = data;
	}

	async* getData(): AsyncIterator<RawSiteMapData[]> {
		if (Array.isArray(this.data)) {
			yield this.data;
		} else {
			yield [ this.data ];
		}
	}
}
