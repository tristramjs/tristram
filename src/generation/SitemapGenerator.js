/* @flow */
import type { Fetcher } from '../fetching';
import type { Writer } from '../persistence';
import type { Formatter } from '../formatting';
import type { RawSiteMapData } from '../types/sitemap';
import chunk from '../util/chunk';

// options management? We have several Modules who needs to know about options
export type Options = {
	hostname: string,
	publicPath?: string,
	cacheTime?: number,
	maxItemsPerSitemap?: number,
	maxItemsPerIndexSitemap?: number,
};

type Props = {
	fetchers: Fetcher[],
	// formatter types..
	formatter: Formatter,
	options: Options,
	writer: Writer,
};

export type OptionsWithDefaults = {
	hostname: string,
	cacheTime?: number,
	maxItemsPerSitemap: number,
	maxItemsPerIndexSitemap: number,
};

type Sitemap = string;

export default class SitemapGenerator {
	fetchers: Fetcher[];
	formatter: Formatter;
	options: OptionsWithDefaults;
	writer: Writer;
	sitemaps: Sitemap[];
	publicPath: string;

	buffer = [];

	constructor({
		fetchers, formatter, options, writer,
	}: Props) {
		this.fetchers = fetchers;
		this.formatter = formatter;
		this.writer = writer;
		this.sitemaps = [];
		this.publicPath = options.publicPath || '';

		this.options = {
			maxItemsPerSitemap: 50000,
			maxItemsPerIndexSitemap: 50000,
			...options,
		};
	}

	async run() {
		for (const fetcher of this.fetchers) {
			for await (const data of fetcher.getData()) {
				await this.saveChunk(data);
			}
		}

		if (this.buffer.length > 0) {
			await this.writeSitemap(this.buffer);
			this.buffer = [];
		}

		await this.writeIndexSitemap();
	}

	async saveChunk(data: RawSiteMapData[]) {
		const { options: { maxItemsPerSitemap } } = this;
		if (this.buffer.length + data.length < maxItemsPerSitemap) {
			this.buffer = this.buffer.concat(data);
		} else if (this.buffer.length + data.length === maxItemsPerSitemap) {
			await this.writeSitemap(this.buffer.concat(data));
			this.buffer = [];
		} else if (this.buffer.length + data.length > maxItemsPerSitemap) {
			// 1. get a chunk of the new data such that (currentItemCount + dataChunk.length == maxItemsPerSitemap)
			// so it can be flushed out to the writer
			const { length: bufferLength } = this.buffer;
			const firstChunk = data.slice(0, maxItemsPerSitemap - bufferLength);
			await this.writeSitemap(this.buffer.concat(firstChunk));
			this.buffer = [];

			// 2. for all chunks where (chunk.length == maxItemsPerSitemap), flush out to Writer
			const [ lastChunk, ...otherChunks ] = chunk(
				data.slice(maxItemsPerSitemap - bufferLength),
				maxItemsPerSitemap
			).reverse();

			for (const part of otherChunks) {
				await this.writeSitemap(part);
			}

			// 3. if the final chunk is (finalChunk.length < maxItemsPerSitemap),
			// create a new sitemap and flush the (incomplete) data chunk ELSE just save the complete sitemap
			if (lastChunk.length < maxItemsPerSitemap) {
				this.buffer = this.buffer.concat(lastChunk);
			} else if (lastChunk.length === maxItemsPerSitemap) {
				await this.writeSitemap(lastChunk);
			}
		}
	}

	async writeSitemap(data: RawSiteMapData[]) {
		const { xmlDeclaration, openingTag, closingTag } = this.formatter;
		const body = this.formatter.format(data);
		const sitemap = [ xmlDeclaration, openingTag, body, closingTag ].join('');

		const map = await this.writer.write(`sitemap-${this.sitemaps.length}.xml`, sitemap);
		this.sitemaps.push(map);
	}

	async writeIndexSitemap() {
		const numberIndexSitemaps = Math.ceil(this.sitemaps.length / this.options.maxItemsPerIndexSitemap);
		const { hostname } = this.options;
		const numberSitemaps = this.sitemaps.length;

		for (let index = 0; index < numberIndexSitemaps; index = index + 1) {
			const indexSitemap = this.formatter.formatIndex({ path: this.publicPath, hostname, numberSitemaps });

			await this.writer.write('index-sitemap.xml', indexSitemap);
		}
	}
}
