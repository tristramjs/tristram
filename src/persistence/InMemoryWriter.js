/* @flow */

import type { Writer } from './index';

export default class InMemoryWriter implements Writer {
	sitemaps = [];

	async createSitemap(xmlDeclaration: string, openingTag: string, partNumber: number) {
		this.sitemaps.push(xmlDeclaration + openingTag);
		return `foo-${partNumber}`;
	}

	async writeChunk(data: string, partNumber: number) {
		this.sitemaps[partNumber] += data;
	}

	async commitSitemap(closingTag: string, partNumber: number) {
		this.sitemaps[partNumber] += closingTag;
	}

	async createIndexSitemap(data: string): Promise<void> {}

	getSitemapPath(partNumber: number) {
		return this.sitemaps[partNumber];
	}
}
