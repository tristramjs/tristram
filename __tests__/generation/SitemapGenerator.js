/* @flow */
import type { Fetcher } from '../../src/fetching/index';
import SitemapGenerator from '../../src/generation/SitemapGenerator';
import SyncFetcher from '../../src/fetching/SyncFetcher';
import PlainFormatter from '../../src/formatting/Plain';
import InMemoryWriter from '../../src/writing/InMemoryWriter';

const fooFetcher: Fetcher<*> = new SyncFetcher({
	data: [ { loc: 'foo' }, { loc: 'bar' }, { loc: 'qak' } ],
});

const barFetcher: Fetcher<*> = new SyncFetcher({
	data: [ { loc: 'Brrrr' } ],
});

const bazFetcher: Fetcher<*> = new SyncFetcher({
	data: [ { loc: 'rrrbbbbB' }, { loc: 'aksjaskjf' } ],
});

const qakFetcher: Fetcher<*> = new SyncFetcher({
	data: [ { loc: 'Rhabarber' }, { loc: 'Rucola' }, { loc: 'Skoda' }, { loc: 'Petersilie' }, { loc: 'Rindsgulasch' } ],
});

describe('SitemapGenerator module', () => {
	[
		{ maxItemsPerSitemap: 2, fetchers: [ fooFetcher, barFetcher, bazFetcher ], result: 4 },
		{ maxItemsPerSitemap: 2, fetchers: [ fooFetcher, bazFetcher ], result: 4 },
		{ maxItemsPerSitemap: 2, fetchers: [ fooFetcher ], result: 3 },
		{ maxItemsPerSitemap: 2, fetchers: [ barFetcher ], result: 2 },
		{ maxItemsPerSitemap: 2, fetchers: [ bazFetcher ], result: 2 },
		{ maxItemsPerSitemap: 2, fetchers: [ qakFetcher, bazFetcher ], result: 5 },
		{ maxItemsPerSitemap: 1, fetchers: [ qakFetcher, bazFetcher ], result: 8 },
	].map(({ maxItemsPerSitemap, fetchers, result }) =>
		it('should run without errors', async () => {
			const options = { hostname: 'foo.bar', maxItemsPerSitemap };

			const writer = new InMemoryWriter();

			const main = new SitemapGenerator({
				fetchers,
				formatter: new PlainFormatter(),
				writer,
				options,
			});

			const sitemaps = await main.run();

			expect(sitemaps).toMatchSnapshot();
			expect(sitemaps.length).toBe(result);
		}));
});
