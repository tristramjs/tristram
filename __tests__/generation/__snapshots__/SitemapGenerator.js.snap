// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SitemapGenerator module should run without errors 1`] = `
Array [
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>foo</loc></url><url><loc>bar</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>qak</loc></url><url><loc>Brrrr</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>rrrbbbbB</loc></url><url><loc>aksjaskjf</loc></url></urlset>",
]
`;

exports[`SitemapGenerator module should run without errors 2`] = `
Array [
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>foo</loc></url><url><loc>bar</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>qak</loc></url><url><loc>rrrbbbbB</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>aksjaskjf</loc></url></urlset>",
]
`;

exports[`SitemapGenerator module should run without errors 3`] = `
Array [
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>foo</loc></url><url><loc>bar</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>qak</loc></url></urlset>",
]
`;

exports[`SitemapGenerator module should run without errors 4`] = `
Array [
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>Brrrr</loc></url></urlset>",
]
`;

exports[`SitemapGenerator module should run without errors 5`] = `
Array [
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>rrrbbbbB</loc></url><url><loc>aksjaskjf</loc></url></urlset>",
]
`;

exports[`SitemapGenerator module should run without errors 6`] = `
Array [
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>Rhabarber</loc></url><url><loc>Rucola</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>Skoda</loc></url><url><loc>Petersilie</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>Rindsgulasch</loc></url><url><loc>rrrbbbbB</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>aksjaskjf</loc></url></urlset>",
]
`;

exports[`SitemapGenerator module should run without errors 7`] = `
Array [
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>Rhabarber</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>Petersilie</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>Skoda</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>Rucola</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>Rindsgulasch</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>rrrbbbbB</loc></url></urlset>",
  "<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>aksjaskjf</loc></url></urlset>",
]
`;
